### DISCUSSION TITLE
macOS Rails app crashes when creating image thumbnail

### DISCUSSION THREAD

**Discussion by sedubois** [libvips/libvips#2077] [Q&A]
**Describe the bug**

In a Rails 6.1 app with ActiveStorage and Vips for image resizing, which calls `resize_to_limit` in the `image_processing` gem, where it seems to call the Vips `thumbnail` command, the transformation does not work when executed on my macbook, but works fine once deployed in production.

**To Reproduce**

- This works on my command line: `vips thumbnail orig.jpg thumb.jpg 1000`
- This works in my Rails console: 
```ruby
> require "image_processing/vips"
> resized = ImageProcessing::Vips.source(file).resize_to_limit!(300, 300)
=> #<File:/var/folders/vs/bktmf9hj0xj9nv1_q7gncjb00000gn/T/image_processing20201208-11617-aqy26k.jpg>
```
- I also ran the code from an ActiveJob, but same result, also works
```ruby
require 'image_processing/vips'

class ImageResizingJob < ApplicationJob
  def perform
    file = File.open("/Users/sdubois/Downloads/vips-test/orig.jpg")
    resized = ImageProcessing::Vips.source(file).resize_to_limit!(800, 800)
    puts "resized: #{resized.inspect}"
  end
end

ImageResizingJob.perform_later
```
- Therefore the only way I can think of to reproduce this would be, in the same environment as I have, to scaffold a Rails app e.g. with a model called Post and with `has_one_attached :image` defined in the Post model, set `Rails.application.config.active_storage.variant_processor = :vips` in an initializer, create a Post and attach an image to it e.g. with `post.image.attach(io: File.open("/path/to/image.jpg"), filename: "image.jpg", content_type: "image/jpg")`, then in some view put something like `post.image.variant(resize_to_limit: [400, 400])`.

**Expected behavior**

No error in Rails console, and the resized image shows up fine in the web app.

**Actual behavior**

This shows up in my Rails log:
```
objc[9725]: +[NSString initialize] may have been in progress in another thread when fork() was called.
objc[9725]: +[NSString initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
```

And the resized image doesn't show up in the web page.

I have been wondering in which repo I should file the issue, as the logs and the workaround point to a low-level issue, but I can only reproduce it in the context of a Rails app.

I am not really blocked by this as I can workaround it in two ways:
- starting Rails with the environment variable `OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES` (found by web searching for the error mentioned in the logs)
- replacing the vips library with the default mini_magick (`Rails.application.config.active_storage.variant_processor = :mini_magick`). 

However I preferred to document the behavior for others and in case there is a way to fix this bug.

**Environment**
 - macOS 11.0.1 
 - ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-darwin20]
 - Rails 6.1.0.rc2 with ActiveStorage and `Rails.application.config.active_storage.variant_processor = :vips`
 - image_processing 1.12.1
 - ruby-vips 2.0.17
 - vips-8.10.2-Mon Oct 12 16:43:59 UTC 2020 installed via homebrew: `brew info vips` => `vips: stable 8.10.2 (bottled)`

**Comment by jcupitt** [comment]
Hi @sedubois, thank you for the report.

I'm not an expert on this, but I think the problem is caused by `vips_init()` being called before the `fork()`.

libvips is highly threaded, and you can't safely have fork and threads in the same process. You need to set up the base environment, then fork to handle a request, and only then init libvips.

I don't think this is really a bug, it's just one of the consequences of using fork, so I'll remove the bug tag.

**Comment by sedubois** [comment]
Thanks @jcupitt, understood. Although from my end-user perspective it does feels like a bug, even if I can't say where it is coming from 😅

**Comment by jcupitt** [comment]
If it's a bug, it's in whatever is starting up libvips before it forks. Whether that's rails or image_processing I couldn't tell you :(

**Comment by sedubois** [comment]
Maybe @janko would have an idea in which part of the stack this might occur?

**Comment by janko** [comment]
I might have experienced this error in the past, but I don't remember now what was the reason or the solution. I think it had to do something with when to require `image_processing/vips` (which requires `vips`), but I don't remember if it was supposed to be sooner or later.

Creating a minimal self-contained Rails app that reproduces the problem would be of big help. It's also important to know which web server are you using locally and on production, and with what configuration, and does the error happen in during requests or background jobs.

I agree that it's fork-safety related and most likely not an issue in libvips.

**Comment by jcupitt** [comment]
ruby-vips does the `vips_init` on require, so one possible fix would be to somehow delay the `require "vips"` until the first request. I've no idea if that's feasible, of course :(

**Comment by sedubois** [comment]
@janko my web server is Puma. I'll see if/when I can assemble a demo but time is currently limited ahead of the holiday season!

**Comment by sedubois** [comment]
Not sure if this can explain but Rails 6.1 changed the way it loads `image_processing`:
https://github.com/rails/rails/commit/74476f0f901e8283cfc2d891e964c5cc6adf8269

**Comment by janko** [comment]
@jcupitt This is difficult to do for forking web servers that allow loading the application before forking (in order to utilize copy-on-write optimizations). In Rails, we also have this issue with libraries that create socket connections, such as database libraries, and we solve this by disconnecting before forking, and let the connection be made again after forking.

I saw that libvips has `vips_shutdown()`, but it explicitly states that you cannot load `vips_init()` again after calling `vips_shutdown()`. If this was possible, that could solve the problem, as we could call `vips_shutdown()` before forking and then again `vips_init()` after, or we could call `vips_shutdown()` & `vips_init()` after forking.

@sedubois Note that this is probably MacOS-specific, which stopped being able to use threads after forking that have been created before forking. You should be able to solve it by disabling Puma's hybrid mode in development, which should prevent it from using forking.

**Comment by jcupitt** [comment]
Huh interesting @janko. I'll have a look at somehow allowing shutdown and restart.

Just to be clear, this sequence is what you need, is that correct?

1. vips_init()
2. vips_shutdown()
3. fork()
4. vips_init()
5. Process some images

ie. no images processed before 5.

**Reply by casperisfine** [reply]
I just stumbled on this because I have a similar need. Was this idea ever experimented with?

For context in our case we're frequently reforking (https://github.com/Shopify/pitchfork/blob/master/docs/REFORKING.md) so delaying `vips_init()` to post fork isn't an option.

**Comment by janko** [comment]
@jcupitt That's correct. Some web servers only have an "after fork" callback, so ideally one could swap steps 2. and 3. and have libvips still work.

**Comment by jcupitt** [comment]
Hello all, I think Kleis might have found the problem (or part of it) -- macOS is missing `sched_getattr` and this forces glib to start threads sooner. 

https://github.com/libvips/php-vips-ext/issues/42#issuecomment-879817757

I've pushed a change to the head of 8.11 to delay thread startup a bit longer, and this might fix this. Would anyone be able to test?

**Reply by terracatta** [reply]
I recently converted from ImageMagick to Vips for my RoR 6.1 project, using `vips-8.11.3-Wed Aug 11 09:29:27 UTC 2021` and the `image_processing` gem, with Puma, and I'm getting `+[NSString initialize] may have been in progress in another thread when fork() was called.` 

So it looks like more is needed to fix this.

**Reply by jcupitt** [reply]
Hi @terracatta, thanks for the report. Is this also on macOS?

**Reply by jcupitt** [reply]
... `[NSString initialize]` looks like an Objective-C-ism (ie. `NextStep String`). Could this error be from your GUI rather than from libvips? We don't use the NSString class.

**Reply by brenogazzola** [reply]
I have just tested this on a fully up to date system (mac, vips, gems) and can confirm it's working. I'm not getting any warnings about `NSString either`. 

@terracatta are you still having problems? `libvips` is the new Rails default, replacing ImageMagick, so I'd like to make sure there's nothing wrong before it gets an official release.

**Reply by matthewford** [reply]
Hi, we're seeing this issue on a M1 Mac running latest Monterey, with vips 8.12

**Reply by jcupitt** [reply]
There have been some new discoveries on this issue (new to me anyway) here:

https://github.com/libvips/ruby-vips/issues/155

With a possible simple fix: add `ruby-vips` to your gemfile directly.
